### PR TITLE
Changes to identify service principal with custom roles on Azure storage account for principal-prefix-access

### LINKED ISSUE #886: [FEATURE]: Identity Service Principal with CustomRole access on storage account `databricks labs ucx save_azure_storage_accounts`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

save_azure_storage_accounts cmd currently looks for SPN with Built in blob roles (Blob Owner, Blob, Contributor and Blog Reader), but doesn't cover Custom Roles with read access on the storage used by clusters. Extend the save_azure_storage_accounts to include custom roles as well

### Proposed Solution

to be defined

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by mohanab-db**
## Changes
1. Changes to identify service principal with custom roles on Azure storage account
2. Unit tests added to increase overall coverage

### Linked issues
Resolves #886 

### Functionality 
- [ ] modified existing command: `databricks labs ucx principal-prefix-access`

### Tests
- [ ] manually tested
- [ ] added unit tests

### CODE REVIEW

**Review by JCZuurmond — COMMENTED** [review]
Hi @mohanab-db, could you please look at my comments?

File: tests/unit/workspace_access/test_scim.py
Comment (nkvuong) [inline tests/unit/workspace_access/test_scim.py:207]: @mohanab-db can you split the additional unit tests into a separate PR? makes it easier to review & revert if necessary

File: src/databricks/labs/ucx/azure/resources.py
Comment (nkvuong) [inline src/databricks/labs/ucx/azure/resources.py:466]: should make this a dataclass instead of a dict

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py:92]: Hi @JCZuurmond Thanks for reviewing the code. I am working on the comments, once done I will create a new PR and share the details with you. 
For this one, I have a question. How the existing functionality for built-in roles work is, if you have 2 role assignments (Blob Reader built-in role and Blob Contributor built-in role) to a service principal (sp1), then the output will be 2 rows -- one row for sp1 read files, one role for sp1 write files.
Similarly, with the new changes, if I have 2 role assignments -- one Blob Contributor built-in role, one read access through custom role, then the output will be 2 rows -- one row for sp1 write files, one role for sp1 read files.
Please let me know if I should handle both the scenarios (existing and new) to reflect only the higher privilege.

File: tests/unit/workspace_access/test_scim.py
Comment (mohanab-db) [reply tests/unit/workspace_access/test_scim.py]: @nkvuong Thanks for reviewing. I have removed additional test from this PR and created another PR just with the unit tests - https://github.com/databrickslabs/ucx/pull/1586

File: src/databricks/labs/ucx/azure/resources.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/resources.py]: @nkvuong - Done. Please check.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py]: @JCZuurmond - Done, please review.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py]: Done, please review.

File: src/databricks/labs/ucx/azure/resources.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/resources.py]: Done, please review.

File: src/databricks/labs/ucx/azure/resources.py
Comment (nkvuong) [inline src/databricks/labs/ucx/azure/resources.py:464]: you can assign directly from the dataclass
```suggestion
            role_name=role_details.role_name,
            role_type=role_details.role_type,
            role_permissions=role_details.role_permissions,
```

File: src/databricks/labs/ucx/azure/access.py
Comment (nkvuong) [inline src/databricks/labs/ucx/azure/access.py:92]:
```suggestion
        for each_permission in role_permissions:
            privilege = self._get_permission_level(each_permission)
            if not privilege:
                continue
            # WRITE_FILES is the higher permission, don't need to check further    
            if privilege == Privilege.WRITE_FILES:
                return privilege
            if privilege == Privilege.READ_FILES:
                higher_privilege = Privilege.READ_FILES                
```

File: src/databricks/labs/ucx/azure/access.py
Comment (nkvuong) [inline src/databricks/labs/ucx/azure/access.py:70]: this one feels fragile to me, what if there are multiple * to match?

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [reply src/databricks/labs/ucx/azure/access.py:99]: Ah, I now see that the existing code indeed creates multiple `StoragePermissionMapping` for the same principal when it is assigned multiple roles. So, let's keep that the same for your code changes.

However, we need to figure out if this goes well. Reading the code in `ServicePrincipalMigration` I don't see deduplication of the storage permissions and the create storage credentials are named after service principals. This creates a collision, with a chance of creating the credential with a lower permission (read instead of write). 

@nkvuong : Could you chime here? Do I read that correctly?

File: src/databricks/labs/ucx/azure/resources.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/resources.py:151]:
```suggestion
    role_permissions: list[str]
```

File: src/databricks/labs/ucx/azure/resources.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/resources.py:144]:
```suggestion
    role_permissions: list[str]
```

File: src/databricks/labs/ucx/azure/resources.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/resources.py:474]:
```suggestion
            role_type = definition_properties.get("type", "BuiltInRole")
```

File: src/databricks/labs/ucx/azure/access.py
Comment (nkvuong) [reply src/databricks/labs/ucx/azure/access.py:99]: that looks correct to me. the risk of collision was lower before as it is less likely for a principal to have both contributor & read access (it is redundant). if we now include custom roles, the collision risk is higher, so I would prefer if the changes are added in this PR

File: src/databricks/labs/ucx/azure/resources.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/resources.py:477]:
```suggestion
                role_permissions_list = definition_properties.get("permissions", [])
```

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [reply src/databricks/labs/ucx/azure/access.py:70]: Is the regex needed? Would a exact match not suffice?

**Review by JCZuurmond — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/azure/resources.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/resources.py]: @nkvuong - Thanks, I have added this.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py]: @nkvuong - Thanks, I have added this.

File: src/databricks/labs/ucx/azure/resources.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/resources.py]: @JCZuurmond - Thanks, I have added this.

File: src/databricks/labs/ucx/azure/resources.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/resources.py]: @JCZuurmond - Thanks, I have added this.

File: src/databricks/labs/ucx/azure/resources.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/resources.py]: @JCZuurmond - Thanks, I have added this.

File: src/databricks/labs/ucx/azure/resources.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/resources.py]: @JCZuurmond - Thanks, I have added this.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py:97]: @JCZuurmond @nkvuong - I have made changes to handle this scenario in the last commit. Please review.

File: src/databricks/labs/ucx/azure/access.py
Comment (nkvuong) [inline src/databricks/labs/ucx/azure/access.py:99]: this mean custom roles take higher priority than pre-defined role, which could be wrong. the cleanest approach is to collate all the output into a set and then emit WRITE_FILES if it is in the set, else emit READ_FILES.

Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py:99]: @nkvuong Thanks for reviewing. One role assignment can either be Custom, or Built-in. If one SP (sp1) has 2 role assignments (one custom read and one built-in write) for a container, then the first iteration in the loop will put READ in the principal_spm_mapping dict. The next iteration will find the write access from built-in and update the dict with the WRITE access.
I have added this scenario to unit test test_save_spn_permissions_custom_role_valid_azure_storage_account() for Custom_role_11. Please review and let me know if this doesn't seem correct.

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/access.py:108]: This could cause a collision when two storage accounts have the containers with the same name. 

``` suggestion
                spm_mapping_key = f"{container}_{role_assignment.principal.client_id}"
```

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/access.py:7]: Learned something new here :)

Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/access.py:7]:
``` suggestion
import re
import uuid
from collections.abc import ValuesView
from dataclasses import dataclass
from functools import partial
```

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/access.py:21]: Also isort these

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/access.py:75]: Comment seems out of place

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/access.py:84]:
```suggestion
            if privilege is None:
```

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/access.py:79]: It might make sense to let this return a list of privileges and select the highest privilege in `_map_storage`. Now the logic for selecting the highest priviliges is duplicated.

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
@mohanab-db : Thanks for implementing our feedback. Functionaly we are getting there, have some remarks about the implementation

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py:108]: @JCZuurmond The _map_storage() method is called for every storage account separately and the dict principal_spm_mapping gets refreshed for every storage account. The dict will only contain container names for one storage account at a time.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py]: @JCZuurmond Thanks. Code updated with this.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py]: @JCZuurmond - Done

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py]: @JCZuurmond  - Done.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py]: @JCZuurmond - Thanks. Changed.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py:79]: @JCZuurmond - The method _get_custom_role_privilege() gets the higher privilege for one custom role assignment. A custom role can contain multiple permissions, and this will get the higher privilege from these permissions. Method _map_storage() will get the higher privilege from a custom or builtin role if more than one role assignments exist for one principal. If I want to return a list of privileges for one custom role, then I will need to iterate through all permissions, whereas now I am exiting from the loop if I find one Write privilege. Please let me know if you think this makes sense.

**Review by JCZuurmond — APPROVED** [review]
LGTM!

File: src/databricks/labs/ucx/azure/access.py
Comment (JCZuurmond) [reply src/databricks/labs/ucx/azure/access.py:79]: yes, that is also true. I am a bit indifferent about this, therefore, I would not let it block this PR. 

Pros: exit loop early, a pattern you see more often in ucx
Cons: treating WRITE_FILES > READ_FILES in two location, speed gain is probably insignifficant

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py:99]: @nkvuong Please let me know if this makes sense, or if some changes should be made.

File: src/databricks/labs/ucx/azure/access.py
Comment (mohanab-db) [reply src/databricks/labs/ucx/azure/access.py]: @nkvuong Please let me know if this makes sense, or if some changes should be made.

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
@mohanab-db : Don't forget to update the README as it now only mentions the built-in roles

**Comment by mohanab-db** [comment]
> @mohanab-db : Don't forget to update the README as it now only mentions the built-in roles

@JCZuurmond - Sorry, I am unable to find the documentation that mentions about built-in roles. I can just see comments about the service principal in general.
image
image
Please point me to the correct documentation so that I can update it.

**Comment by JCZuurmond** [comment]
Hi @mohanab-db, I saw it being mentioned in the `migrate-credentials` command. I think we should move this information to the `principal-prefix-access` command and reference the SP in `migrate-credentials` to be the ones created by `principal-prefix-access`

**Comment by mohanab-db** [comment]
@JCZuurmond - I have updated README.md. Please review.

**Review by nkvuong — APPROVED** [review]

**Review by JCZuurmond — APPROVED** [review]
lgtm

**Pull request merged**
